fedimint-client/src/sm/executor.rs

```diff
@@ -312,115 +349,133 @@ where
 
-        let num_states = active_states.len();
-        let num_transitions = transitions.len();
-        debug!(
-            num_states,
-            num_transitions, "Awaiting any state transition to become ready"
-        );
-
-        let ((transition_outcome, state, transition_fn, meta), _, _) = select! {
-            res = select_all(transitions) => res,
-            () = new_state_added => {
-                debug!("New state added, re-starting state transitions");
+        loop {
+            if active_state_count == 0 {
+                debug!(
+                    "No state transitions remaining, exiting execute_next_state_transitions loops"
+                );
                 return Ok(());
             }
-        };
-
-        debug!(
-            ?state,
-            transition_outcome = ?AbbreviateJson(&transition_outcome),
-            "Executing state transition"
-        );
-
-        let new_state = self
-            .db
-            .autocommit(
-                |dbtx| {
-                    let state = state.clone();
-                    let transition_fn = transition_fn.clone();
-                    let transition_outcome = transition_outcome.clone();
-                    Box::pin(async move {
-                        let new_state = transition_fn(
-                            &mut ClientSMDatabaseTransaction::new(dbtx, state.module_instance_id()),
-                            transition_outcome,
-                            state.clone(),
-                        )
-                        .await;
-                        dbtx.remove_entry(&ActiveStateKey::from_state(state.clone()))
-                            .await;
-                        dbtx.insert_entry(
-                            &InactiveStateKey::from_state(state.clone()),
-                            &meta.into_inactive(),
-                        )
-                        .await;
-
-                        let context = &self
-                            .module_contexts
-                            .get(&state.module_instance_id())
-                            .expect("Unknown module");
-
-                        let global_context =
-                            global_context_gen(state.module_instance_id(), state.operation_id());
-                        if new_state.is_terminal(context, &global_context) {
-                            // TODO: log state machine id or something
-                            debug!("State machine reached terminal state");
-                            dbtx.insert_entry(
-                                &InactiveStateKey::from_state(new_state.clone()),
-                                &ActiveState::new().into_inactive(),
+            let num_states = active_state_count;
+            let num_transitions = transitions.len();
+            debug!(
+                num_states,
+                num_transitions, "Awaiting any state transition to become ready"
+            );
+            let new_state_added = async move {
```

Comment:
We still break the loop if new states are added and all existing futures are pending. Perhaps we could query the active states and then figure out what states are new in some way. Then we could just add them to the list of transitions.

@elsirion @m1sterc001guy  ideas here?

Reply:
Refactoring this into a channel would be good, I think I wanted to avoid forgetting a place where states are created when I built this and didn't have the time for a full refactor.

In the channel version we'd only load from DB on startup and then only add states via the channel. We'd still save the state to DB before sending it over the channel for crash safety though.